### PR TITLE
Alternative Zenith fix (Skyline check)

### LINKED ISSUE #3987: zenith visual beam is hide-able, thus NOT revealing position
the visual beam into the sky of zenith is not visible if you park something in its path, e.g. charon

http://zero-k.info/Battles/Detail/916987

this is a strategic advantage, hiding the position when you know this. is this intended? else fix or give the skybeam damage? -.-

### PR DESCRIPTION
**Pull request by Arch-Shaman**
Fixes #3987 by implementing a "skyline" check through GetUnitsInCylinder and GetFeaturesInCylinder. This is an alternative solution to PR #3990.

### CODE REVIEW

**Comment by Arch-Shaman** [comment]
Also tried Spring.GetUnitWeaponHaveFreeLineOfFire(unitID,1, ux,uy,uz,ux,9000,uz) -- this doesn't seem to provide the right results so I opted for getunitsincylinder with an additional getfeatures with 20 radius (+4 elmo from beam radius) which visually seemed to provide the correct behavior. The waterweapon addition to the beam is necessary otherwise the player can simply do UW guass nonsense with it and block the beam that way.

**Comment by GoogleFrog** [comment]
I would much rather this be done registering and handling the event of the Zenith beam impacting something before fading out, as this seems like it would be computationally free in normal usage. The other PR (making the beam penetrate everything) is fine though.

**Comment by Arch-Shaman** [comment]
> I would much rather this be done registering and handling the event of the Zenith beam impacting something before fading out, as this seems like it would be computationally free in normal usage.

ExplosionCreated or WatchWeapon is the closest thing I could think of. Unless there's something else I'm not seeing?

**Comment by sprunk** [comment]
> ExplosionCreated or WatchWeapon is the closest thing I could think of. Unless there's something else I'm not seeing?

IIRC fading lasers don't trigger the Explosion callin, but ProjectileDestroyed should be fair game (called both on expiry and collision, but just an `y > 5000` check should be fine). WatchWeapon just enables the other callins and will need to be used regardless.

**Comment by Arch-Shaman** [comment]
> IIRC fading lasers don't trigger the Explosion callin, but ProjectileDestroyed should be fair game (called both on expiry and collision, but just an y > 5000 check should be fine). WatchWeapon just enables the other callins and will need to be used regardless.

I'll check it out. Thanks for the info.

**Comment by GoogleFrog** [comment]
I would start by looking at Explosion. You don't need to detect the case where the projectile fades out, so watching for ProjectileDestroyed should not be required.

**Comment by Arch-Shaman** [comment]
Should be good to go barring any unforeseen things.

File: LuaRules/Gadgets/unit_zenithskylinecheck.lua [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  }
+end	
+
+
+local gravityWeaponDefID = WeaponDefNames["zenith_gravity_neg"].id
+local spSetUnitRulesParam = Spring.SetUnitRulesParam
+
+function gadget:Explosion(weaponDefID, px, py, pz, AttackerID, ProjectileID)
+	if weaponDefID == gravityWeaponDefID then
+		spSetUnitRulesParam(AttackerID, "isBlocked", 1)
```
Comment (sprunk) [inline LuaRules/Gadgets/unit_zenithskylinecheck.lua:23]: "isBlocked" is a bit generic, how about "meteorSpawnBlocked"

File: LuaRules/Gadgets/unit_zenithskylinecheck.lua [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	end
+end
+
+function gadget:Initialize()
+	Script.SetWatchWeapon(gravityWeaponDefID, true)
```
Comment (sprunk) [inline LuaRules/Gadgets/unit_zenithskylinecheck.lua:28]: IIRC there might be a more specific SetWatchWeaponExplosion

File: scripts/zenith.lua [hunk 3]
```diff
@@ -191,8 +192,10 @@ local function SpawnProjectileThread()
 			RegainControlOfMeteors()
 			currentlyStunned = false
 		end
-		
-		SpawnMeteor()
+		if not spGetUnitRulesParam(unitID,"isBlocked") or spGetUnitRulesParam(unitID,"isBlocked") == 0 then -- Only spawn if there's a clear view to the sky
```
Comment (sprunk) [inline scripts/zenith.lua:195]: could do `~= 1` to handle both `nil` and `0`

File: units/zenith.lua [hunk 4]
```diff
@@ -105,6 +105,7 @@ return { zenith = {
       turret                  = true,
       weaponType              = [[LaserCannon]],
       weaponVelocity          = 6000,
+      waterWeapon             = true,
```
Comment (sprunk) [inline units/zenith.lua:108]: Should be brought back to false now that the visual actually matches the logic

**Comment by sprunk** [comment]
> You don't need to detect the case where the projectile fades out

Ideally the fade would be what triggers the meteor to spawn. Currently the block is just applied to the next beam (which is not a problem given the reloadtime is both not high enough for this to be noticeable and not low enough for there to be 2 beams inbound).

**Comment by sprunk** [comment]
Also I just realized a minor issue: you can prevent the "meteors past the cap fall on your head" this way. IIRC that was done by design to incentivize usage (but it's a bit opaque on what is happening, and other incentives already exist, so this is mostly just the cool factor in the end).

File: units/zenith.lua (hunk 4 above)
Comment (Arch-Shaman) [reply units/zenith.lua:108]: I think if it's false, the check won't go through and the meteors will gather anyways. I can run a test to see if this is true though.

File: scripts/zenith.lua (hunk 3 above)
Comment (Arch-Shaman) [reply scripts/zenith.lua:195]: Good point, I'll update it.

File: units/zenith.lua (hunk 4 above)
Comment (sprunk) [reply units/zenith.lua:108]: If it doesn't call `Explode` on hitting water that's an engine bug. Or maybe it needs a fake water CEG defined to do that.

Comment (Arch-Shaman) [reply units/zenith.lua:108]: It won't even fire underwater. It's similar to if you take a ravager and drop it in the sea AFAIK.

**Comment by Arch-Shaman** [comment]
> Also I just realized a minor issue: you can prevent the "meteors past the cap fall on your head" this way. IIRC that was done by design to incentivize usage (but it's a bit opaque on what is happening, and other incentives already exist, so this is mostly just the cool factor in the end).

What if being blocked for some time caused the meteors to fall.

**Comment by GoogleFrog** [comment]
I think being blocked should be equivalent to being turned off - the meteors instantly fall. Falling meteors are recaptured when the Zenith turns back on with the current on/off toggle so we need not worry about a plane flying over and destroying your base.

The change is essentially a bit of rule-of-cool for Raven or Krow interactions.

**Comment by Arch-Shaman** [comment]
Ravens while testing were extremely finicky to get to block the beam (as was licho) but Krow is probably more reasonable interaction expectation. I'll add it to the to do.

**Comment by GoogleFrog** [comment]
Are you sure this tries to fire but is blocked underwater? I don't see how that could be the case with waterWeapon = true.

**Comment by Arch-Shaman** [comment]
> Are you sure this tries to fire but is blocked underwater?

With waterweapon=false, the zenith fires once (when initially spawned) then never fires again: https://streamable.com/izm9mk

**Pull request merged**